Injected workspace peer dependency error (7.0.0-rc.2 - strict-peer-dependencies)

### pnpm version:

7.0.0-rc.2

### Code to reproduce the issue:

https://github.com/PabloSzx/injected-workspace-peer-pnpm


```json
{
  "name": "a",
  "version": "0.0.1",
  "devDependencies": {
    "b": "workspace:^0.0.1"
  },
  "peerDependencies": {
    "b": "workspace:^0.0.1"
  }
}
```

```json
{
  "name": "b",
  "version": "0.0.1"
}
```

```json
{
  "name": "c",
  "version": "0.0.1",
  "dependencies": {
    "a": "workspace:^0.0.1"
  },
  "devDependencies": {
    "b": "workspace:^0.0.1"
  },
  "dependenciesMeta": {
    "a": {
      "injected": true
    }
  }
}
```

```sh
pnpm -v
7.0.0-rc.2
```

```sh
pnpm update -r
Scope: all 3 workspace projects
 ERR_PNPM_PEER_DEP_ISSUES  Unmet peer dependencies

packages/c
└─┬ a
  └── ✕ unmet peer b@workspace:^0.0.1: found 0.0.1
```


### Expected behavior:

To ignore the "workspace:" for the peer dependency version validation

### Actual behavior:

Breaks because of the "workspace:" syntax

### Additional information:

I had to do `pnpm add -D b --strict-peer-dependencies=false` just to be able to add "b" as dev dep

 - `node -v` prints: v16.14.0
 - Windows, macOS, or Linux?: macOS

--- a/.changeset/small-snakes-think.md
+++ b/.changeset/small-snakes-think.md
@@ -1,0 +1,7 @@
+---
+"@pnpm/core": patch
+"@pnpm/resolve-dependencies": patch
+"pnpm": patch
+---
+
+Peer dependency should be correctly resolved from the workspace, when it is declared using a workspace protocol [#4529](https://github.com/pnpm/pnpm/issues/4529).

--- a/packages/core/test/install/injectLocalPackages.ts
+++ b/packages/core/test/install/injectLocalPackages.ts
@@ -1202,3 +1202,93 @@ test('inject local packages when node-linker is hoisted and dependenciesMeta is 
     expect(modulesState?.injectedDeps?.['project-1'][1]).toEqual(path.join('project-3', 'node_modules', 'project-1'))
   }
 })
+
+test('peer dependency of injected project should be resolved correctly', async () => {
+  const project1Manifest = {
+    name: 'project-1',
+    version: '1.0.0',
+    dependencies: {},
+  }
+  const project2Manifest = {
+    name: 'project-2',
+    version: '1.0.0',
+    devDependencies: {
+      'project-1': 'workspace:1.0.0',
+    },
+    peerDependencies: {
+      'project-1': 'workspace:^1.0.0',
+    },
+  }
+  const project3Manifest = {
+    name: 'project-3',
+    version: '1.0.0',
+    dependencies: {
+      'project-1': 'workspace:1.0.0',
+      'project-2': 'workspace:1.0.0',
+    },
+    dependenciesMeta: {
+      'project-2': {
+        injected: true,
+      },
+    },
+  }
+  preparePackages([
+    {
+      location: 'project-1',
+      package: project1Manifest,
+    },
+    {
+      location: 'project-2',
+      package: project2Manifest,
+    },
+    {
+      location: 'project-3',
+      package: project3Manifest,
+    },
+  ])
+
+  const importers: MutatedProject[] = [
+    {
+      buildIndex: 0,
+      manifest: project1Manifest,
+      mutation: 'install',
+      rootDir: path.resolve('project-1'),
+    },
+    {
+      buildIndex: 0,
+      manifest: project2Manifest,
+      mutation: 'install',
+      rootDir: path.resolve('project-2'),
+    },
+    {
+      buildIndex: 0,
+      manifest: project3Manifest,
+      mutation: 'install',
+      rootDir: path.resolve('project-3'),
+    },
+  ]
+  const workspacePackages = {
+    'project-1': {
+      '1.0.0': {
+        dir: path.resolve('project-1'),
+        manifest: project1Manifest,
+      },
+    },
+    'project-2': {
+      '1.0.0': {
+        dir: path.resolve('project-2'),
+        manifest: project2Manifest,
+      },
+    },
+    'project-3': {
+      '1.0.0': {
+        dir: path.resolve('project-3'),
+        manifest: project2Manifest,
+      },
+    },
+  }
+  await mutateModules(importers, await testDefaults({
+    nodeLinker: 'hoisted',
+    workspacePackages,
+  }))
+})

--- a/packages/core/test/install/injectLocalPackages.ts
+++ b/packages/core/test/install/injectLocalPackages.ts
@@ -1291,4 +1291,8 @@ test('peer dependency of injected project should be resolved correctly', async (
     nodeLinker: 'hoisted',
     workspacePackages,
   }))
+
+  const rootModules = assertProject(process.cwd())
+  const lockfile = await rootModules.readLockfile()
+  expect(lockfile.packages?.['file:project-2_project-1@project-1'].dependencies?.['project-1']).toEqual('link:project-1')
 })

--- a/packages/resolve-dependencies/src/resolvePeers.ts
+++ b/packages/resolve-dependencies/src/resolvePeers.ts
@@ -399,7 +399,7 @@ function resolvePeers<T extends PartialResolvedPackage> (
   const resolvedPeers: {[alias: string]: string} = {}
   const missingPeers = []
   for (const peerName in ctx.resolvedPackage.peerDependencies) { // eslint-disable-line:forin
-    const peerVersionRange = ctx.resolvedPackage.peerDependencies[peerName]
+    const peerVersionRange = ctx.resolvedPackage.peerDependencies[peerName].replace(/^workspace:/, '')
 
     const resolved = ctx.parentPkgs[peerName]
     const optionalPeer = ctx.resolvedPackage.peerDependenciesMeta?.[peerName]?.optional === true

